### PR TITLE
twisterlib: refine test cases collection logic

### LINKED ISSUE #45845: tests: The failure test case number increase significantly in CMSIS DSP tests on ARM boards.
**Describe the bug**
After the twister overhaul PR https://github.com/zephyrproject-rtos/zephyr/pull/44094, it reveal some incorrect testcase numbers in the twister report. 
Take frdm_k64f as an example, following tests are affected most:  
tests/lib/cmsis_dsp/martix  => 5940   failure
tests/lib/cmsis_dsp/transform  => 1100  failure
tests/lib/cmsis_dsp/filtering => 2172 failure

Because that testcases cannot be parsed well when twister to scan the test case from source code. We did not see this issue because
these incorrect testcases was judge as "skipped" instead of "failure" before this PR, but issues existed before.

**To Reproduce**
1. source zephry-env.sh
2. twister -p frdm_k64f --device-testing --device-serial=/dev/ttyACM0  -T tests/lib/cmsis_dsp/ -v

**Expected behavior**
The failure testcases should be very less.

**Impact**
The failure testcases increase significantly in the test report.

**Logs and console output**
```
<testsuites><testsuite name="frdm_k64f" timestamp="2022-05-18T01:48:09.248298" time="115.16999999999999" tests="15000" failures="9282" errors="0" skipped="154"><properties>
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain (e.g Zephyr SDK 0.14.2)
 - Commit SHA: d78194546746

### PR DESCRIPTION
**Pull request by smrtos**
Mark unintended test cases as skipped in the test plan.

Due to the diverse coding styles and lack of preprocessing when
scanning for test cases, there are many unintended combinations
of yaml test suites and C test functions in the preliminary test
plan. This patch tries to identify such unintended ones by checking
the function existence in the zephyr.lst file. And mark them as
"skipped in C source code".

To make the amending accurate, it is done right after the build
stage of the pipeline and before runtime case collection happens.
For those test instances that don't go through the build stage,
such as those with "Skip filtered" reason, they are not affected
by this amending.

Signed-off-by: Ming Shao <ming.shao@intel.com>

Fix #45845

### CODE REVIEW

**Comment by smrtos** [comment]
Tried with below command:
twister --device-testing --device-serial /dev/ttyACM0 --device-serial-baud 115200 -p reel_board -T tests/lib/cmsis_dsp/matrix/ -vv

Without this patch: the passrate is 8.33%

With this patch: the pass rate is 100%

**Comment by smrtos** [comment]
This pr may fix: https://github.com/zephyrproject-rtos/zephyr/issues/45845

**Comment by hakehuang** [comment]
this is not a good solution, if some cases was controlled by marco and we just missing that marco, the fucntion will be stripped by compiler, and we will miss it. Besides, we add many dependency on the tools,  and why not fix it at the scan function directly?

**Comment by smrtos** [comment]
> ...if some cases was controlled by marco and we just missing that marco, the fucntion will be stripped by compiler, and we will miss it...

Current situation is, the initial test plan has many incorrect combinations of yaml test scenarios and test functions. 

Incorrect ones like:
`libraries.cmsis_dsp.matrix.unary_q7.arm_mat_inverse_f32`

Correct ones like:
`libraries.cmsis_dsp.matrix.unary_q7.op1_arm_mat_trans_q7`

Test function `arm_mat_inverse_f32` will never be part of the binary compiled for `libraries.cmsis_dsp.matrix.unary_q7`.
So it shouldn't be part of the test plan.  But currently twister doesn't have a mechanism to remove such false combinations.

> ...Besides, we add many dependency on the tools...
I changed to use `zephyr.lst` now and remove the dependency on the `nm` tool.
This should also make it workable on Windows.

> ...and why not fix it at the scan function directly?
If we want to collect the intended cases precisely with static scanning,  we may have to do a proper C preprocessing first.
That's too much effort I think.

**Comment by smrtos** [comment]
Fix a few typos.

**Comment by hakehuang** [comment]
> If we want to collect the intended cases precisely just by static scanning, we may have to do a proper C preprocessing first.
> That's too much effort I think.

ok, then this could be a solution, thanks for this PR.

**Review by hakehuang — DISMISSED** [review]

**Review by enjiamai — DISMISSED** [review]

**Comment by enjiamai** [comment]
This will fix https://github.com/zephyrproject-rtos/zephyr/issues/45845.

**Comment by smrtos** [comment]
Add more logic in the Test harness to try to identify test cases from non-ztest-compliant tests.
As I tried, it revealed some test cases which we never realized before.
Such as this one: 
https://github.com/zephyrproject-rtos/zephyr/blob/fa2054f93e4e80b079a4a6a9e84d642e51912042/tests/arch/arm/arm_runtime_nmi/src/arm_runtime_nmi.c#L26

**Comment by enjiamai** [comment]
> https://github.com/zephyrproject-rtos/zephyr/blob/fa2054f93e4e80b079a4a6a9e84d642e51912042/tests/arch/arm/arm_runtime_nmi/src/arm_runtime_nmi.c#L26
Yes, @smrtos , this testcases always report pass currently, see: https://github.com/zephyrproject-rtos/zephyr/pull/45998 for more detail.

**Comment by smrtos** [comment]
> 

I think we don't need https://github.com/zephyrproject-rtos/zephyr/pull/45998.
This patch can handle that case.

**Comment by smrtos** [comment]
@hakehuang @enjiamai I made more changes about Test harness. Could you take a look? Thanks!

**Review by hakehuang — DISMISSED** [review]

File: scripts/pylib/twister/harness.py [hunk 1]
```diff
@@ -285,8 +291,28 @@ def handle(self, line):
             self._match = False
             self.ztest = True
 
+        # if a PASS/FAIL/SKIP log line doesn't contain "test_" prefix,
+        # it is considered as a non-ztest-comliant test case.
+        # but it may still output log to indicate PASS/FAIL/SKIP.
+        # such test cases shouldn't be silently swallowed.
+        # so try to extract the test case status from its log line
+        if match and not match.group(2):
```
[the remark this answers is not in the document]
Comment (smrtos) [reply scripts/pylib/twister/harness.py:299]: Below case will output: `PASS - nmi_test_isr`. It has no `test_` prefix.
https://github.com/zephyrproject-rtos/zephyr/blob/fa2054f93e4e80b079a4a6a9e84d642e51912042/tests/arch/arm/arm_runtime_nmi/src/arm_runtime_nmi.c#L26

So it satisfies `match and not match.group(2)`.

This is a non-ztest-compliant test but well formed.

And this PR make it detectable from log.

[the remark this answers is not in the document]
Comment (LixinGuoX) [reply scripts/pylib/twister/harness.py:299]: Yes, it makes sense for that circumstance, and I think this test has some problem in logic, maybe we should improve that too.

**Comment by gopiotr** [comment]
First of all - thank you @smrtos for your contribution and for your proposition to solving problem with detection of Ztest testcases from C files.

So far I analyzed only this part connected with harness modifications. I have some doubts connected with introduction this verification of `non-ztest-compliant` test cases. I don't think so, that this approach can solve a problem with tests which are not written in Ztest framework. Let's consider this test:
`zephyr/tests/arch/arm/arm_runtime_nmi`
With your external `non-ztest-compliant` test cases detection new testcase is detected during executing this line:
https://github.com/zephyrproject-rtos/zephyr/blob/fa2054f93e4e80b079a4a6a9e84d642e51912042/tests/arch/arm/arm_runtime_nmi/src/arm_runtime_nmi.c#L25
But this is not any new testcase... this test fragment is a part of already existed test case defined here:
https://github.com/zephyrproject-rtos/zephyr/blob/fa2054f93e4e80b079a4a6a9e84d642e51912042/tests/arch/arm/arm_runtime_nmi/src/main.c#L14

So, with your changes I can observe in JSON report following two testcases:
```json
{
    "identifier":"arch.interrupt.arm.nmi.arm_runtime_nmi",
    "execution_time":"10.10",
    "status":"passed"
},
{
    "identifier":"arch.interrupt.arm.nmi.nmi_test_isr",
    "execution_time":"10.10",
    "status":"passed"
},
```

But in real there are no two separate testcases. `arch.interrupt.arm.nmi.nmi_test_isr` was performed inside `arch.interrupt.arm.nmi.arm_runtime_nmi` and shouldn't be considered as a separate testcase.

My opinion is that, we should fix the tests source code, rather than try to fit Twister to every possible case. Especially that there are only few tests which use macros like `TC_END_RESULT` or `TC_END_REPORT`:
`zephyr/tests/arch/arm/arm_runtime_nmi`
`zephyr/tests/crypto/tinycrypt`
`zephyr/tests/drivers/ipm`

I appreciate your idea, but in my opinion this solution complicate harness too much and in result it do not improve significantly test detection. 

Moreover, I don't think so, that decision about if testcase is `ztest-compliant` or `non-ztest-compliant` should based only on existence or not `test_` prefix in test case name. This is not robustness solution - there is no limit for defining test in `TC_START` macro by name starting from `test_` like for example:
`TC_START("test_examplary_name")`
In such situation your code will recognize wrongly such test case as a ZTest test case.

To summarize:
I propose to leave in this PR harness modification and solving problem with tests based only on `TC_START`, `TC_END_RESULT` and `TC_END_REPORT` macros and focus on only this one writing with explicitly defined testcases by `ztest_unit_test()` or `ZTEST()`. 
In my opinion problem with tests based only on `TC_START`, `TC_END_RESULT` and `TC_END_REPORT` macros should be solved by PRs like this: https://github.com/zephyrproject-rtos/zephyr/pull/45998.

File: scripts/pylib/twister/twisterlib.py [hunk 2]
```diff
@@ -2767,6 +2770,49 @@ def process(self, pipeline, done, message, lock, results):
...
+                # by checking the log, such test cases may not have functions in .lst file.
+                # Also remove them here, which should be OK because they will still be added back at runtime.
+                if not (test_func_name1 in lst_content or test_func_name2 in lst_content):
+                    logger.debug(f"Unintended test case: {identifier}")
+                    continue
+
+                intended_testcases.append(testcase)
+                logger.debug(f"Intended test case: {identifier}")
+
+        self.instance.testcases.clear()
```
[the remark this answers is not in the document]
Comment (smrtos) [reply scripts/pylib/twister/twisterlib.py:2813]: But such combinations are totally irrational. They are included in the test plan merely because we didn't do a proper C preprocessing. If we have a chance to suppress them, why not?

File: scripts/pylib/twister/twisterlib.py [hunk 3]
```diff
@@ -2767,6 +2770,49 @@ def process(self, pipeline, done, message, lock, results):
...
+    def amend_test_instance(self):
+        lst_file = os.path.join(self.build_dir, "zephyr", "zephyr.lst")
+        if os.path.isfile(lst_file):
+            logger.debug(f"zephyr.lst located: {lst_file}")
+        else:
+            # no zephyr.lst file, cannot do symbol-based amending
+            logger.debug(f"zephyr.lst NOT located: {lst_file}")
+            return
+
+        yaml_testsuite_name = self.instance.name.split(os.path.sep)[-1]
```
[the remark this answers is not in the document]
Comment (smrtos) [reply scripts/pylib/twister/twisterlib.py:2782]: Thanks! That's much better.

File: scripts/pylib/twister/twisterlib.py [hunk 4]
```diff
@@ -2767,6 +2770,49 @@ def process(self, pipeline, done, message, lock, results):
...
+            lst_content = fp.read()
+            logger.debug(f"test instance {self.instance.name} has {len(self.instance.testcases)} cases.")
+            for testcase in self.instance.testcases:
+                # identifier is "yaml_testsuite.test_function" or just "yaml_testsuite"
+                identifier = testcase.name
+                if identifier == yaml_testsuite_name:
+                    # Sometimes twsiter add yaml test suite name to a test instance as a test case.
+                    # Such test case has no real test function.
+                    # Skip it to make stats right
+                    logger.debug(f"skipping: {identifier}")
```
[the remark this answers is not in the document]
Comment (smrtos) [reply scripts/pylib/twister/twisterlib.py:2796]: Let me try it tmr and get back.

File: scripts/pylib/twister/twisterlib.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gopiotr) [reply scripts/pylib/twister/twisterlib.py:2813]: I agree with you that test plan are prepared in wrong way, due to the fact that we are not able to preprocess C files better. And generating test plan after run tests (after run building) is not the best solution - the idea of test plan is to create "plan" with information what test will be performed, to make it possible to verify at the end if all planed tests was really performed (or eventually why not). As far as I understand, the test plan should help us detect some kind of "silent" skips performed during runtime.

If you completely remove information from final report about this unintended test cases, then you get huge discrepancy between test plan and test report. And potential user can be confused - what happen with those missing testcases?

To solve this problem entirely the tests source code should be changed - to be possible to parse them properly by Twister. But I know, that this is too much to do at this moment. So, for me some kind of compromise is marking those unintended tests as "skip" - this is still better solution rather than marking them as "null".

File: scripts/pylib/twister/harness.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nashif) [reply scripts/pylib/twister/harness.py:299]: this test has bogus output, we should not change twister for this, the test itself needs to be fixed. (The test should not have `TC_START("nmi_test_isr");`, which should come from ztest.

File: scripts/pylib/twister/twisterlib.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (smrtos) [reply scripts/pylib/twister/twisterlib.py:2813]: @PerMac Yes, the `testcases` keyword is the most reliable way so far. It bypasses the source file parsing and offloads the duty to the case author. If we can change all the existing cases as necessary and ensure coding discipline, this can be a solution. This PR is meant to mitigate the incorrect result caused by source file parsing.

**Comment by smrtos** [comment]
@hakehuang @gopiotr  Could you help review the latest code? Thanks!

**Review by gopiotr — DISMISSED** [review]

**Review by hakehuang — DISMISSED** [review]

**Review by chen-png — DISMISSED** [review]

**Review by nashif — CHANGES_REQUESTED** [review]
still confused what is this fixing that is not already fixed. Please show the issues this PR is fixing using some example calls to twister..

**Comment by hakehuang** [comment]
> still confused what is this fixing that is not already fixed. Please show the issues this PR is fixing using some example calls to twister..

@nashif this issue is about tset cases in cmsis_dsp where the each .c files defines the ZTEST_SUITE, and the test suite has its own test cases by DEFINE_TEST_VARIANT5 in its own files, but our twister script will scan the whole folder and match test suite with all DEFINE_TEST_VARIANT5, as in cmsis_dsp the valid cases for suite are only defined in file scope not in folder scope. this will case there are many error cases in arm platform test report. @smrtos would you report one issue for your PR, you fix can helps to resolve the unwantted errors cases in test report, which is a block issue for us to upload test report.

**Comment by smrtos** [comment]
> still confused what is this fixing that is not already fixed. Please show the issues this PR is fixing using some example calls to twister..

@nashif 

With below command:

`./scripts/twister --device-testing --device-serial /dev/ttyACM0 --device-serial-baud 115200 -p reel_board -vv -T tests/lib/cmsis_dsp/matrix`


Test case like below one will be collected by twister:

`"identifier":"libraries.cmsis_dsp.matrix.unary_q7.op1_arm_mat_trans_f64"`

This is an invalid combination because the yaml test suite `libraries.cmsis_dsp.matrix.unary_q7` will never choose the test function `test_op1_arm_mat_trans_f64` for its config macros.

1. With the old twister as I submitted the PR, such test case will have result like this:

```
      "identifier":"libraries.cmsis_dsp.matrix.unary_q7.op1_arm_mat_trans_f64",
      "execution_time":"0.00",
      "status":"blocked",
      "reason":"Unknown"
```

2. With the latest twister (after merging https://github.com/zephyrproject-rtos/zephyr/pull/46278), such test case will have result like this:

```
      "identifier":"libraries.cmsis_dsp.matrix.unary_q7.op1_arm_mat_trans_f64",
      "execution_time":"0.00",
      "status":null
```

3. Suppose this PR is merged, the result will look like this:

```
      "identifier":"libraries.cmsis_dsp.matrix.unary_q7.op1_arm_mat_trans_f64",
      "execution_time":"0.00",
      "status":"skipped",
      "reason":"Skipped in test C source code"
```

I created this PR in order to fix the incorrect combination of yaml test suite and the scanned test functions.
Though https://github.com/zephyrproject-rtos/zephyr/pull/46278 addressed this issue in a different way (see result 2), 
I think result 3 is more reasonable and less intimidating to the user.

BTW, if you just try this PR alone, you may find some failures like below:

```
        {
            "identifier":"libraries.cmsis_dsp.basicmath",
            "execution_time":"0.00",
            "status":"blocked",
            "reason":"Unknown"
        },
```

This is because the old twister will add a test case named after the yaml test suite if no test functions are scanned from c file.
This PR doesn't try to fix that. It is fixed by https://github.com/zephyrproject-rtos/zephyr/pull/46278 at：
https://github.com/zephyrproject-rtos/zephyr/blob/26294ced096cb8a5491ffe826b675df2af6aab1f/scripts/pylib/twister/twisterlib.py#L2073.

So, if the community agrees, I suggest we merge this PR to give user a more reasonable test result.

**Comment by smrtos** [comment]
> > 3. With the latest twister (after merging [twister: fix setting status on unprocessed tests #46278](https://github.com/zephyrproject-rtos/zephyr/pull/46278)), such test case has result like this:
> > ```
> >       "identifier":"libraries.cmsis_dsp.matrix.unary_q7.op1_arm_mat_trans_f64",
> >       "execution_time":"0.00",
> >       "status":null
> > ```
> > 
> you are showing the json data here, however, in the final xml report those tests are being marked as skipped already

I compared the final xml as below. Maybe it is inappropriate/intimidating to mark such test suite as "misconfiguration".
And if we don't amend them immediately after the build stage, we may not be able to tell them from those really mis-configured ones. @nashif 

1. The latest twister output this for "libraries.cmsis_dsp.matrix.unary_q7.op1_arm_mat_trans_f64"
```
<testcase classname="libraries.cmsis_dsp" name="libraries.cmsis_dsp.matrix.unary_q7.op1_arm_mat_trans_f64" time="0.00">
<skipped type="untested" message="No results captured, testsuite misconfiguration?"/>
</testcase>
```

2. With this PR, the output is like this:
```
<testcase classname="libraries.cmsis_dsp" name="libraries.cmsis_dsp.matrix.unary_q7.op1_arm_mat_trans_f64" time="0">
<skipped type="skipped" message="Skipped in test C source code"/>
</testcase>
```

**Comment by nashif** [comment]
> I compared the final xml as below. Such cases are not marked as skipped. But rather as "untested/misconfiguration". Maybe it is inappropriate/intimidating to do so.

but this is exactly what is happening. They are skipped with a different reason that can help fix the actual testsuite and not simply work around it.

**Comment by smrtos** [comment]
> > I compared the final xml as below. Such cases are not marked as skipped. But rather as "untested/misconfiguration". Maybe it is inappropriate/intimidating to do so.
> 
> but this is exactly what is happening. They are skipped with a different reason that can help fix the actual testsuite and not simply work around it.

TBH, I don't want to work around it.  With this PR, we can tell such cases from other untested cases and give an exact reason. If user wants fix it, he/she can still fix it. And with more clues. 

Anyway, the bottom line is to make the final stats right. And https://github.com/zephyrproject-rtos/zephyr/pull/46278 is fairly enough for that purpose.

File: scripts/pylib/twister/twisterlib.py [hunk 5]
```diff
@@ -2767,6 +2770,49 @@ def process(self, pipeline, done, message, lock, results):
...
+                    continue
+
+                test_func_name1 = f"test_{identifier.split('.')[-1]}"
+                test_func_name2 = f"{identifier.split('.')[-1]}"
```
[the remark this answers is not in the document]
Comment (PerMac) [reply scripts/pylib/twister/twisterlib.py:2798]: > That's because I'm afraid someone may not follow the function name convention. So just in case. :)

This shouldn't take place. We should enforce following the conventions and not adding "sloppiness". And in fact we already are:
https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib.py#L2005

File: scripts/pylib/twister/twisterlib.py [hunk 6]
```diff
@@ -2767,6 +2770,43 @@ def process(self, pipeline, done, message, lock, results):
...
+                    # Skip it to make stats right
+                    continue
+
+                test_func_name1 = f"test_{identifier.split('.')[-1]}"
+                test_func_name2 = f"{identifier.split('.')[-1]}"
+                # If the .lst file exists and the test case function not in it,
+                # it should be an un-intended test function for this test instance build.
+                # If this is an old test plan and contains test cases collected at runtime
+                # by checking the log, such test cases may not have functions in .lst file.
+                # Also remove them here, which should be OK because they will still be added back at runtime.
```
Comment (PerMac) [inline scripts/pylib/twister/twisterlib.py:2803]: I believe these comments are not accurate anymore, right? Can you fix them?

File: scripts/pylib/twister/twisterlib.py [hunk 7]
```diff
@@ -2767,6 +2770,43 @@ def process(self, pipeline, done, message, lock, results):
...
+                if not (test_func_name1 in lst_content or test_func_name2 in lst_content):
+                    testcase.status = "skipped"
+                    testcase.reason = "Skipped in test C source code"
```
Comment (PerMac) [inline scripts/pylib/twister/twisterlib.py:2806]: IMO "Skipped in test C source code" sounds like something that was intended, but in fact happens due to issues when parsing C code. As I understand, our end goal would be to eliminate such cases, so we have precise test plan before test execution. By improving parser, fixing test cases, etc. or adding `testcases` entry directly to yamls if nothing else was possible. 
I suggest using something like "Skipped due to an inaccurate test plan" to show that there is still sth to be improved.

**Review by PerMac — CHANGES_REQUESTED** [review]
I like this idea of parsing a `lst` file a lot. I saw @nashif approach which was merged, however I believe this approach is more "direct". I think there can be several scenarios which can end with inaccurate test plan and "untested/misconfiguration" test cases. This PR adds a clear and direct detection of (biggest?) part of such cases at their root.

I added some minor issues that I think have to be fixed yet.

@smrtos What do you think about adding a check, that when `testcases` entry is present in the yaml file, `amend_test_instance` is skipped? I think it is possible that directly given `testcases` might end wrong, and `amend_test_instance` will shadow it pointing to issues with C code. Such cases would get `misconfiguration` status from Anas PR.  

Can you also check how it affects twister performance-wise? I guess it is simple parsing of small files, so I don't expect much difference, but would like to get some confirmation.

File: scripts/pylib/twister/twisterlib.py (hunk 6 above)
Comment (smrtos) [reply scripts/pylib/twister/twisterlib.py:2803]: Thanks. It is fixed now.

**Comment by smrtos** [comment]
> @smrtos What do you think about adding a check, that when `testcases` entry is present in the yaml file, `amend_test_instance` is skipped? I think it is possible that directly given `testcases` might end wrong, and `amend_test_instance` will shadow it pointing to issues with C code. Such cases would get `misconfiguration` status from Anas PR.

It is added now.


> Can you also check how it affects twister performance-wise? I guess it is simple parsing of small files, so I don't expect much difference, but would like to get some confirmation.

I did compare the performance. I cannot recall the exact time cost but the impact is very small.

File: scripts/pylib/twister/twisterlib.py (hunk 7 above)
Comment (PerMac) [reply scripts/pylib/twister/twisterlib.py]: TBH I don't really agree :) For me  "Skipped in test C source code" would apply to cases, when there is an explicit command to skip, like due to #ifdef or some decorator in the code etc. However, if it is due to some convoluted macro as in the case of cmsis_dsp then IMO it is not really "skipped in C **source code**". Anyhow, I see your point and this is just some minor wording disagreement :)

**Review by PerMac — APPROVED** [review]

**Comment by nashif** [comment]
Still not convinced this is the right approach, although it might be heading in the right direction. The issue here is that with complex tests we end up a test list we expect that is never run, i.e. because the tests are never compiled due to build time configuration. Parsing the c file has been the way for us to discover tests which we know can be problematic with some autogenerated tests, something that will continue to have..

So, I am thinking we should look into using this approach or a similar approach (maybe use output of nm instead of looking at the .lst file) to generate the test plan instead of the very awkward regex based parsing we do now. This would depend on the build step to be completed, but this way we can capture all tests that are supposed to run for each scenario accurately.

With the new ztest API we will get all the information we need, including the testsuite name, which is another gap we have now. For example, for tests/kernel/common (when converted to new API):

```
nm build/zephyr/zephyr.elf | grep z_ztest_unit_test_
000000000011b4a8 d z_ztest_unit_test_atomic_tests_test_atomic
000000000011b4c8 d z_ztest_unit_test_atomic_tests_test_threads_access_atomic
000000000011b4e8 d z_ztest_unit_test_bitarray_tests_test_bitarray_alloc_free
000000000011b508 d z_ztest_unit_test_bitarray_tests_test_bitarray_declare
000000000011b528 d z_ztest_unit_test_bitarray_tests_test_bitarray_region_set_clear
000000000011b548 d z_ztest_unit_test_bitarray_tests_test_bitarray_set_clear
000000000011b568 d z_ztest_unit_test_bitarray_tests_test_ffs
000000000011b588 d z_ztest_unit_test_bitfield_tests_test_bitfield
000000000011b5a8 d z_ztest_unit_test_boot_delay_tests_test_bootdelay
000000000011b5c8 d z_ztest_unit_test_bound_check_tests_test_bounds_check_mitigation
000000000011b608 d z_ztest_unit_test_byteorder_tests_test_byteorder_memcpy_swap
000000000011b5e8 d z_ztest_unit_test_byteorder_tests_test_byteorder_mem_swap
000000000011b628 d z_ztest_unit_test_byteorder_tests_test_sys_get_be16
000000000011b648 d z_ztest_unit_test_byteorder_tests_test_sys_get_be24
000000000011b668 d z_ztest_unit_test_byteorder_tests_test_sys_get_be32
000000000011b688 d z_ztest_unit_test_byteorder_tests_test_sys_get_be48
000000000011b6a8 d z_ztest_unit_test_byteorder_tests_test_sys_get_be64
000000000011b6c8 d z_ztest_unit_test_byteorder_tests_test_sys_get_le16
000000000011b6e8 d z_ztest_unit_test_byteorder_tests_test_sys_get_le24
000000000011b708 d z_ztest_unit_test_byteorder_tests_test_sys_get_le32
000000000011b728 d z_ztest_unit_test_byteorder_tests_test_sys_get_le48
000000000011b748 d z_ztest_unit_test_byteorder_tests_test_sys_get_le64
000000000011b768 d z_ztest_unit_test_byteorder_tests_test_sys_put_be16
000000000011b788 d z_ztest_unit_test_byteorder_tests_test_sys_put_be24
000000000011b7a8 d z_ztest_unit_test_byteorder_tests_test_sys_put_be32
000000000011b7c8 d z_ztest_unit_test_byteorder_tests_test_sys_put_be48
000000000011b7e8 d z_ztest_unit_test_byteorder_tests_test_sys_put_be64
000000000011b808 d z_ztest_unit_test_byteorder_tests_test_sys_put_le16
000000000011b828 d z_ztest_unit_test_byteorder_tests_test_sys_put_le24
000000000011b848 d z_ztest_unit_test_byteorder_tests_test_sys_put_le32
000000000011b868 d z_ztest_unit_test_byteorder_tests_test_sys_put_le48
000000000011b888 d z_ztest_unit_test_byteorder_tests_test_sys_put_le64
000000000011b8a8 d z_ztest_unit_test_clock_tests_test_clock_cycle_32
000000000011b8c8 d z_ztest_unit_test_clock_tests_test_clock_cycle_64
000000000011b8e8 d z_ztest_unit_test_clock_tests_test_clock_uptime
000000000011b908 d z_ztest_unit_test_clock_tests_test_ms_time_duration
000000000011b928 d z_ztest_unit_test_errno_tests_test_errno
000000000011b948 d z_ztest_unit_test_errno_tests_test_thread_context
000000000011b968 d z_ztest_unit_test_irq_offload_tests_test_irq_offload
000000000011b988 d z_ztest_unit_test_irq_offload_tests_test_nested_irq_offload
000000000011b9a8 d z_ztest_unit_test_multilib_tests_test_multilib
000000000011b9c8 d z_ztest_unit_test_pow2_tests_test_pow2_ceil
000000000011b9e8 d z_ztest_unit_test_printk_tests_test_printk
000000000011ba08 d z_ztest_unit_test_timeout_order_tests_test_timeout_order
000000000011ba28 d z_ztest_unit_test_version_tests_test_version
```

This is exactly the list of tests with the testsuite that can be easily parse and can populate the testplan.

Doing so would solve all issues and we will finally have a testplan that is accurate and reports that are not full of skips because of bad parsing and auto-generated tests...

**Comment by smrtos** [comment]
@nashif 

> So, I am thinking we should look into using this approach or a similar approach (maybe use output of nm instead of looking at the .lst file) to generate the test plan instead of the very awkward regex based parsing we do now. This would depend on the build step to be completed, but this way we can capture all tests that are supposed to run for each scenario accurately.

At first I did use the `nm` tool.  But as @hakehuang suggested, it may introduce extra dependency on the Linux-specific toolchain. Twister may be supposed to run on Windows as well. So I changed to use the plain `.lst` file, which offers equivalent info but only needs plain file read.

> With the new ztest API we will get all the information we need, including the testsuite name, which is another gap we have now. For example, for tests/kernel/common (when converted to new API):
> 
> ```
> nm build/zephyr/zephyr.elf | grep z_ztest_unit_test_
> ...

I am not quite sure about the new ztest API yet. But it looks promising. Hope it can provide all the clues about an accurate test plan.

**Comment by nashif** [comment]
> it may introduce extra dependency on the Linux-specific toolchain.

nm is not linux specific, it exists on all host supported by the SDK. We can still use .lst, but with the output of nm it is more straightforward.

> I am not quite sure about the new ztest API yet. But it looks promising. Hope it can provide all the clues about an accurate test plan.

all tests need to move this new API, so this should be our target.

**Review by hakehuang — APPROVED** [review]

**Comment by nashif** [comment]
> @nashif Is it OK for me to remove all the regex-based test function collection and just rely on the symbols from elf or .lst file? That's less twisted indeed.

not in the first stage I guess. 
I would prefer to have this come on top of the restructring/refactor PR, otherwise we will be in rebase hell.

**Comment by smrtos** [comment]
> > @nashif Is it OK for me to remove all the regex-based test function collection and just rely on the symbols from elf or .lst file? That's less twisted indeed.
> 
> not in the first stage I guess. I would prefer to have this come on top of the restructring/refactor PR, otherwise we will be in rebase hell.

Sure. I will just put the restructring/refactor PR here: https://github.com/zephyrproject-rtos/zephyr/pull/46528

**Comment by PerMac** [comment]
@smrtos I wouldn't remove parser based test cases collection until we have another fully operational mean. The lst/nm check requires an image to get built, so it doesn't help for the test plan creation, which has to be fast. However is still very useful as a validation of created test plan. 

My idea of approach to incorrect test cases in test plan: 
- we merge this PR
- we create the list (and monitor) of all results getting "Skipped due to an inaccurate test plan"
- those can be the tests to be fixed first. Either by porting them to new ztest api (if it solves the uncertainties in the test plan) or just manually add `testcases` entry to their yamls.
- After no more "inaccurate test plan" are observe we change this check to produce error instead of skip. This will then help us to prevent regressions (tests not aligned with our standards) from entering the main

@nashif @smrtos What do you think about such approach?

**Comment by nashif** [comment]
> @smrtos I wouldn't remove parser based test cases collection until we have another fully operational mean of getting test cases in advance. The lst/nm check requires an image to get built, so it doesn't help for the test plan creation, which has to be fast. However is still very useful as a validation of created test plan.

Right, this is not going to be fast and would need everything to be built first, the main disadvantage, reason why I want to leave the parsing logic and add this as another mean for getting the correct list of test cases for mostly problematic tests. Might still be considered as an option and if we decide to go for accuracy over speed.

> My idea of approach to incorrect test cases in test plan:
> 
> * we merge this PR
> * we create the list (and monitor) of all results getting "Skipped due to an inaccurate test plan"
> * those can be the tests to be fixed first. Either by porting them to new ztest api (if it solves the uncertainties in the test plan) or just manually add `testcases` entry to their yamls.
> * After no more "inaccurate test plan" are observe we change this check to produce error instead of skip. This will then help us to prevent regressions (tests not aligned with our standards) from entering the main
> 
> @nashif @smrtos What do you think about such approach?

I do not agree with this approach. This PR in the current states and on its own does not solve any problems IMO. My other concern here is that we barking the wrong tree and nothing will be fixed in the actual tests.
Fixing tests directly or addressing test case collection in problematic tests should be really our main priority, combined with the approach taken in this PR.

Listing test cases in yaml files works fine for basic tests and can be manageable, but it does not work for complex tests and is very cumbersome to maintain when you have 20 scenarios with 100s of tests cases each. This is the reason it was not used for example for the cmsis dsp tests.

This is what I am proposing:
- Remove parsing complexity by moving to new ztest API ASAP
- While migrating to new ztest API, fix most test cases that have complex logic, new ztest API should help with that.
- This PR should work on generating the test plan and not as a post step to verify the test plan and set status. We could make this work on-demand by adding an option to the yaml file to generate list of test cases only on select test suites while we continue to parse things as before on everything else.

**Comment by nashif** [comment]
@smrtos  are you going to rework this?

**Comment by smrtos** [comment]
@nashif  yes, I will do that.

**Comment by smrtos** [comment]
This change will be tracked with: https://github.com/zephyrproject-rtos/zephyr/pull/49619.
Closing this one.
